plugin/trino-sqlserver/src/test/java/io/trino/plugin/sqlserver/BaseSqlServerTypeMapping.java

```diff
@@ -50,6 +171,155 @@ public void testVarbinary()
+                .addRoundTrip("nvarchar(5)", "N'攻殻機動隊'", createVarcharType(5), "CAST('攻殻機動隊' AS varchar(5))")
+                .addRoundTrip("nvarchar(32)", "N'攻殻機動隊'", createVarcharType(32), "CAST('攻殻機動隊' AS varchar(32))")
+                .addRoundTrip("nvarchar(20)", "N'😂'", createVarcharType(20), "CAST('😂' AS varchar(20))")
+                .addRoundTrip("nvarchar(77)", "N'Ну, погоди!'", createVarcharType(77), "CAST('Ну, погоди!' AS varchar(77))")
+                .execute(getQueryRunner(), onRemoteCreateAndInsert("test_varchar"));
+
+        SqlDataTypeTest.create()
+                .addRoundTrip("varchar(10)", "'text_a'", createVarcharType(10), "CAST('text_a' AS varchar(10))")
+                .addRoundTrip("varchar(255)", "'text_b'", createVarcharType(255), "CAST('text_b' AS varchar(255))")
+                .addRoundTrip("varchar(5)", "CAST('攻殻機動隊' AS varchar(5))", createVarcharType(5), "CAST('攻殻機動隊' AS varchar(5))")
+                .addRoundTrip("varchar(32)", "CAST('攻殻機動隊' AS varchar(32))", createVarcharType(32), "CAST('攻殻機動隊' AS varchar(32))")
+                .addRoundTrip("varchar(20)", "CAST('😂' AS varchar(20))", createVarcharType(20), "CAST('😂' AS varchar(20))")
+                .addRoundTrip("varchar(77)", "CAST('Ну, погоди!' AS varchar(77))", createVarcharType(77), "CAST('Ну, погоди!' AS varchar(77))")
+                .execute(getQueryRunner(), trinoCreateAndInsert(getSession(), "test_varchar"))
+                .execute(getQueryRunner(), trinoCreateAsSelect("test_varchar"));
+    }
+
+    @Test
+    public void testVarcharUnbounded()
+    {
+        SqlDataTypeTest.create()
+                .addRoundTrip("varchar", "'text_a'", createUnboundedVarcharType(), "CAST('text_a' AS varchar)")
+                .addRoundTrip("varchar", "'text_b'", createUnboundedVarcharType(), "CAST('text_b' AS varchar)")
+                .addRoundTrip("varchar", "'text_d'", createUnboundedVarcharType(), "CAST('text_d' AS varchar)")
+                .addRoundTrip("varchar", "CAST('攻殻機動隊' AS varchar)", createUnboundedVarcharType(), "CAST('攻殻機動隊' AS varchar)")
+                .addRoundTrip("varchar", "CAST('攻殻機動隊' AS varchar)", createUnboundedVarcharType(), "CAST('攻殻機動隊' AS varchar)")
+                .addRoundTrip("varchar", "CAST('攻殻機動隊' AS varchar)", createUnboundedVarcharType(), "CAST('攻殻機動隊' AS varchar)")
+                .addRoundTrip("varchar", "CAST('😂' AS varchar)", createUnboundedVarcharType(), "CAST('😂' AS varchar)")
+                .addRoundTrip("varchar", "CAST('Ну, погоди!' AS varchar)", createUnboundedVarcharType(), "CAST('Ну, погоди!' AS varchar)")
+                .addRoundTrip("varchar", "'text_f'", createUnboundedVarcharType(), "CAST('text_f' AS varchar)")
+                .execute(getQueryRunner(), trinoCreateAndInsert(getSession(), "test_varchar_unbounded"))
+                .execute(getQueryRunner(), trinoCreateAsSelect("test_varchar_unbounded"));
+    }
+
+    @Test
+    public void testVarcharMapping()
```

Comment:
That's not a perfect method name (everything here is a "test mapping").

You can group the test cases differently

- testSqlServerVarchar
- testSqlServerNvarchar
- testTrinoVarchar
   - this would cover regular case, unbounded, as well as >4000